recorderproxy/https.go

```diff
@@ -0,0 +1,295 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package recorderproxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	//TLSConfig   func(host string, ctx *recordContext) (*tls.Config, error)
+	TLSConfig   = TLSConfigFromCA()
+	httpsRegexp = regexp.MustCompile(`^https:\/\/`)
+)
+
+func stripPort(s string) string {
+	ix := strings.IndexRune(s, ':')
+	if ix == -1 {
+		return s
+	}
+	return s[:ix]
+}
+
+func (proxy *RecorderProxy) dial(network, addr string) (c net.Conn, err error) {
+	if proxy.RoundTripper.Dial != nil {
+		return proxy.RoundTripper.Dial(network, addr)
+	}
+	return net.Dial(network, addr)
+}
+
+func (proxy *RecorderProxy) connectDial(network, addr string) (c net.Conn, err error) {
+	if proxy.ConnectDial == nil {
+		return proxy.dial(network, addr)
+	}
+	return proxy.ConnectDial(network, addr)
+}
+
+func (proxy *RecorderProxy) handleHttps(w http.ResponseWriter, r *http.Request) {
+	ctx := NewRecordContext(proxy)
+
+	hij, ok := w.(http.Hijacker)
+	if !ok {
+		panic("httpserver does not support hijacking")
+	}
+
+	proxyClient, _, e := hij.Hijack()
+	if e != nil {
+		panic("Cannot hijack connection " + e.Error())
+	}
+
+	host := r.URL.Host
+
+	proxyClient.Write([]byte("HTTP/1.0 200 OK\r\n\r\n"))
+	ctx.Logf("Assuming CONNECT is TLS, mitm proxying it")
+	// this goes in a separate goroutine, so that the net/http server won't think we're
+	// still handling the request even after hijacking the connection. Those HTTP CONNECT
+	// request can take forever, and the server will be stuck when "closed".
+	// TODO: Allow Server.Close() mechanism to shut down this connection as nicely as possible
+	tlsConfig := defaultTLSConfig
+	if TLSConfig != nil {
+		var err error
+		tlsConfig, err = TLSConfig(host, ctx)
+		if err != nil {
+			httpError(proxyClient, ctx, err)
+			return
+		}
+	}
+
+	go func() {
+		//TODO: cache connections to the remote website
+		rawClientTls := tls.Server(proxyClient, tlsConfig)
+
+		if err := rawClientTls.Handshake(); err != nil {
+			ctx.Resp = NewResponse(r, "text/plain", 555, "CONNECT failed")
+			ctx.Warnf("Cannot handshake client %v %v", r.Host, err)
+			if ctx.Resp != nil {
+				if err := ctx.Resp.Write(proxyClient); err != nil {
+					ctx.Warnf("Cannot write response that reject http CONNECT: %v", err)
+				}
+			}
+			proxyClient.Close()
+
+			return
+		}
+		defer rawClientTls.Close()
+		clientTlsReader := bufio.NewReader(rawClientTls)
+		for !isEof(clientTlsReader) {
+			req, err := http.ReadRequest(clientTlsReader)
```

Comment:
Make the body of the loop a function to make it more readable and avoid `defer` in a loop.